wolfcrypt/src/port/xilinx/xil-versal-trng.c

```diff
@@ -0,0 +1,221 @@
+        nonce_written += to_write;
+    } while (nonce_written < sizeof(user_cfg->PersString));
+    user_cfg->PersStrPresent = XTRNGPSV_TRUE;
+}
+#endif
+
+/**
+   Initialize the Versal TRNG.
+
+   nonce	Pointer to user-supplied nonce.
+   nonceSz	Size of the user-supplied nonce.
+
+   Returns 0 on success
+ */
+int wc_VersalTrngInit(byte* nonce, word32 nonceSz)
+{
+#if !defined(HAVE_HASHDRBG)
+    /* Use the HRNG mode (with DF) which lets us extract FIPS compliant
+     * random data from the TRNG.
+     */
+    XTrngpsv_UsrCfg user_cfg = {
+            .Mode = XTRNGPSV_HRNG,
+            .SeedLife = WC_RESEED_INTERVAL,
+            .PredResistanceEn = XTRNGPSV_FALSE,
+            .DFDisable = XTRNGPSV_FALSE,
+            .DFLenMul = WC_XILINX_TRNG_DF_LEN_MUL,
+            .InitSeedPresent = XTRNGPSV_FALSE,
+            .PersStrPresent = XTRNGPSV_FALSE
+    };
+#else
+    /* Use the PTRNG mode (without DF) which lets us extract raw entropy
+     * to be used as a seed to the internal DRBG.
+     */
+    const XTrngpsv_UsrCfg user_cfg = {
+            .Mode = XTRNGPSV_PTRNG,
+            .SeedLife = 0U,
+            .PredResistanceEn = XTRNGPSV_FALSE,
+            .DFDisable = XTRNGPSV_TRUE,
+            .DFLenMul = 0U,
+            .InitSeedPresent =  XTRNGPSV_FALSE,
+            .PersStrPresent = XTRNGPSV_FALSE
+    };
+#endif
+    int ret = WC_HW_E;
+    XTrngpsv_Config *cfg;
+    s32 xret = 0;
+    if (trng.State == XTRNGPSV_HEALTHY) {
+        return 0;
+    }
+    cfg = XTrngpsv_LookupConfig(WOLFSSL_PSV_TRNG_DEV_ID);
+    if (!cfg) {
+        WOLFSSL_MSG("Could not lookup TRNG config");
+        goto out;
+    }
+    xret = XTrngpsv_CfgInitialize(&trng, cfg, cfg->BaseAddress);
+    if (xret)
+        goto out;
+    xret = versal_trng_selftest();
+    if (xret)
+        goto out;
+#if !defined(HAVE_HASHDRBG)
+    if (nonce)
+        usercfg_add_nonce(&user_cfg, nonce, nonceSz);
+#endif
+    xret = XTrngpsv_Instantiate(&trng, &user_cfg);
+    if (xret)
+        goto out;
+
+    ret = 0;
+
+out:
+    if (xret) {
+        WOLFSSL_MSG_EX("Xilinx API returned 0x%08x", xret);
+    }
+    return ret;
+}
+
+/**
+   Reset the Versal TRNG.
+
+   Returns 0 on success
+ */
+int wc_VersalTrngReset(void)
+{
+    s32 xret = XTrngpsv_Uninstantiate(&trng);
+    if (xret) {
+        WOLFSSL_MSG_EX("XTrngpsv_Uninstantiate() returned 0x%08x", xret);
+        return WC_HW_E;
+    }
+    return 0;
+}
+
+/**
+   Run the Versal TRNG self-test.
+
+   Returns 0 on success
+ */
+int wc_VersalTrngSelftest(void)
+{
+    return versal_trng_selftest() == XTRNGPSV_SUCCESS ? 0 : -1;
+}
+
+/**
+   Read from the Versal TRNG.
+
+   output	Destination
+   sz		Length of output
+
+   Returns 0 on success
+ */
+int wc_VersalTrngGenerate(byte* output, word32 sz)
+{
+    /* The TRNG always generates exactly 32bytes of output */
+    byte buf[XTRNGPSV_SEC_STRENGTH_BYTES];
+    word32 bytes_generated = 0;
+    do {
+        u32 bytes_left = sz - bytes_generated;
```

Comment:
Using a mix of `word32` and `u32` and `int` and `s32` here. Would prefer ours.

Reply:
My preference is use ours... word32 and int.

Reply:
Done